src/extensions/idle.rs

```diff
@@ -193,8 +193,8 @@ impl<'a, T: Read + Write + 'a> Handle<'a, T> {
         }
     }
 
-    /// Block until the given callback returns `Stop`, or until an unhandled
-    /// response arrives on the IDLE channel.
+    /// Block until the given callback returns `Stop`, or until a response
+    /// arrives that is not explicitly handled by [`UnsolicitedResponse`].
```

Comment:
Actually, here's a crazy though — maybe `WaitOutcome` should have a variant `Unexpected` and _that_ would hold the `imap_proto::Response` that was unexpected.

Reply:
I tried this, but quickly bumped into the reality that the `imap_proto::Response` is referring to data that was read off the wire in `wait_inner`. :)

So we'd have to make sure the stuff coming off the wire lives long enough to be returned to the `wait_*` caller. I guess the simplest way would be to have the caller pass in the buffer, or return a tuple `Unexpected(Vec<u8>, imap_proto::Response<'a>)`. Unless there is an owned version of `Response` that I do not see??

I get the impression you want the user to have the `imap_proto::Response` somehow. Is this so they can match on it, or do we just want them to know 'I saw this'. If we want them to match on it and do something with it, we could just give it to them in the callback instead of an `UnexpectedResponse`. The only reason we're using `UnexpectedResponse` here is because previous versions were sending it down the `unsolicited_responses` channel. Since we're not doing that anymore, we aren't actually bound to it, and it would make this diff a lot simpler. I am not sure I like the disconnect between getting an `imap_proto` type in the callback and `imap` types everywhere else, but maybe we want to see where that goes?

Reply:
You know, I think I just straight-up forgot to add an `into_owned` method for `imap_proto::Response` in https://github.com/djc/tokio-imap/pull/115. That was an oversight, and could trivially be added here:
https://github.com/djc/tokio-imap/blob/bb553c9512393c1a040e19194f95595958f0cb32/imap-proto/src/types.rs#L46

and would look roughly like this:
https://github.com/djc/tokio-imap/blob/bb553c9512393c1a040e19194f95595958f0cb32/imap-proto/src/types.rs#L102

May be worth just submitting a PR there and then we could give out `imap_proto::Response<'static>`!

Well, so, I want the user to be handed an `imap_proto::Response` _only_ if that response doesn't fit into what we expect should be possible responses. You can sort of think of it as a "recoverable error". Basically, in my head, `WaitOutcome` is sort of like a `Result<Timeout, imap_proto::Response>` where the `Response` is yielded if a response was encountered that doesn't fit into `UnsolicitedResponse`, and thus _shouldn't_ be possible.

> I am not sure I like the disconnect between getting an `imap_proto` type in the callback and `imap` types everywhere else, but maybe we want to see where that goes?

Ah, no, that's not what I'm proposing. I'm proposing that the callback continues to get `UnsolicitedResponse`, but that `wait` returns an `imap_proto::Response` _if_ it sees an unexpected response (i.e., one that should be able to come unsolicited).

Reply:
Having `into_owned` on `imap_proto::Result` would solve a lot of these kinds of problems! :)

I can have a stab at this on the weekend? Unless you want to have a go.

Having unexpected responses returned sounds reasonable. So that's all good then. Do we want to hold this PR open until the owned `Response` thing gets sorted, or add it in a later PR?

Reply:
So I went to do this, and ... we *already* return unhandled `Response`s in the `Err` path. Specifically, as `Err(Error::Parse(ParseError::Unexpected(format!("{:?}", err))))` (from [here](https://github.com/jonhoo/rust-imap/blob/4108917c58c4fe5083418b00b6a337e8d3b04ece/src/error.rs#L115)). We're stringifying the unexpected `Response` instead of giving an owned version back to the caller (but we could make that a non-stringy version with an owned `Response` if we wanted).

And now it seems weird to instead be returning `Ok(WaitOutcome::Unexpected(Response))` when this is actually really more of an error condition (ie. this *shouldn't happen*), and we're already doing it as an `Err`.

Thoughts?

Reply:
Oh, nice! Yeah, let's just return the owned response right there. Although maybe not as a `ParseError`? That feels a little odd. Though I understand why it used to be under there.

Reply:
PR for owned `Response` is [here](https://github.com/djc/tokio-imap/pull/120). Had to do `MailboxDatum` at the same time. :)

Am I correct that the reason this is under `Error::ParseError` is because it seems to mostly (entirely?) come out of the various `parse_*` functions. So if we make  this just `Error::Unexpected` I think the `ParseError::Unexpected` variant will never be constructed and can be removed?

It makes sense to me to separate "I could not parse this thing" from "I parsed it but do not know what to do with it", so moving `Unexpected` from a `ParseError` variant to an `Error` variant makes sense to me. Also makes sense if we want to keep it as a `ParseError` variant because "I was parsing and something went wrong". Semantics are fun! :)

Reply:
Yes, that's exactly right. We also used to roll our own parsing, and those errors got wrapped up in `ParseError`, but now that we use `imap_proto` that's also a little different. I _think_ a top-level `Unexpected` variant is the way to go here. And if we then have no more `ParseError` that make sense, then I suppose we should get rid of it! The `enum` is `#[non_exhaustive]` anyway, so we could always add it back in later.

Reply:
Just updating - I have something for this queued up, but waiting on the imap-proto change to make it into a release. I asked today when that might be, so MTF.

Reply:
Pushed an update for this change just now.